data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/criteria/QueryCriteriaMethodMatch.java

```diff
@@ -233,14 +233,7 @@ protected MethodMatchInfo build(MethodMatchContext matchContext) {
             for (QueryModel.Criterion criterion : junction.getCriteria()) {
                 countQuery.add(criterion);
             }
-            // Joins are skipped for count query for OneToMany, ManyToMany
-            // however skipping joins from criteria could cause issues (in many to many?)
             for (JoinPath joinPath : queryModel.getJoinPaths()) {
-                Association association = joinPath.getAssociation();
```

Comment:
The fix for original issue was to remove joins for count query, but that worked only because we had tests with criteria on root entity. Once there are criteria on joined entity, query compilation fails.

So, reverting this will allow query to be compiled and executed, but total size count will not return correct numbers and it's not an easy thing to fix when there are joins. If there is one author with two books, count on author (joined with books) will return 2. And we must add join when there is criteria. The original issue will be reopened.